app/temporal/src/activities.ts

```diff
@@ -81,175 +88,150 @@ export function makeActivities(
+      const firstRecord = await database.mongoClient
+        .db(params.bucket)
+        .collection<DataSetRow>("sourceData")
+        .findOne();
       // all rows should have all available headers (see source file processing)
-      const sourceColumns = pull(Object.keys(jsonData[0]), "__rowId");
+      const sourceColumns = pull(
+        Object.keys(firstRecord as DataSetRow),
+        "__sourceRowId"
+      );
       return dataAnalyzer.generateMappingRecommendations(
         sourceColumns,
         params.columnConfig
       );
     },
     processDataValidations: async (params: {
-      bucket: string;
-      fileReference: string;
+      importerId: string;
       validatorColumns: ColumnValidators;
-      outputFileReference: string;
       stats: SourceFileStatsPerColumn;
-      patches: DataSetPatch[];
-    }): Promise<{ errorFileReference: string; errorCount: number }> => {
-      console.time("validations");
-      const referenceId = params.fileReference.split("-")[1].split(".")[0];
-      const fileData = await fileStore.getFile(
-        params.bucket,
-        params.fileReference
-      );
-      const jsonData = JSON.parse(fileData.toString());
-      const patchedData = applyPatches(jsonData, params.patches);
-      const errorData = dataAnalyzer.processDataValidations(
-        patchedData,
+      skip: number;
+      limit: number;
+    }): Promise<number> => {
+      const jsonData: DataSet = await database.mongoClient
+        .db(params.importerId)
+        .collection<DataSetRow>("data")
+        .find()
+        .skip(params.skip)
+        .limit(params.limit)
+        .toArray();
+
+      const validationResults = dataAnalyzer.processDataValidations(
+        jsonData,
         params.validatorColumns,
         params.stats
       );
-      await fileStore.putFile(
-        params.bucket,
-        params.outputFileReference,
-        Buffer.from(JSON.stringify(errorData))
-      );
-      console.timeEnd("validations");
-      return {
-        errorFileReference: params.outputFileReference,
-        errorCount: errorData.length,
-      };
+
+      for (const validationResult of validationResults) {
+        await database.mongoClient
+          .db(params.importerId)
+          .collection("data")
+          .updateOne(
+            { __sourceRowId: validationResult.rowId },
+            {
+              $push: {
+                [`data.${validationResult.column}.messages`]:
+                  validationResult.messages,
+              },
+            }
+          );
+      }
+
+      return validationResults.length;
+    },
+    applyPatches: async (params: {
+      importerId: string;
+      patches: DataSetPatch[];
+    }): Promise<void> => {
+      for (const patch of params.patches) {
+        const targetColumn = `data.${patch.column}`;
+        await database.mongoClient
+          .db(params.importerId)
+          .collection("data")
+          .updateOne(
+            { __sourceRowId: patch.rowId },
```

Comment:
Warum nicht die _id?

Reply:
können wir auch - bisher haben wir in den patches die rowId gespeichert, wenn wir die _id speichern wollen, dann müssen wir den Datentyp auf ObjectId ändern.